File: ups_server/cmd_arg.c
/* 
 * Modul cmd_arg definuje funkce pro zpracování argumentů programu.
 * 
 * Author: Petr Kozler
 */

#include "cmd_arg.h"
#include "config.h"
#include "string_utils.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <stdbool.h>
#include <netinet/in.h>

/**
 * Zpracuje předané argumenty příkazové řádky.
 * 
 * @param argc počet argumentů příkazové řádky
 * @param argv argumenty příkazové řádky (IP adresa a port pro naslouchání)
 * @return návratová hodnota programu
 */
args_t parse_args(int argc, char** argv) {
    if (argc <= 1) {
        usage();
    }
    
    args_t args;
    args.host = DEFAULT_HOST;
    bool has_port = false;
    
    int i = 1;
    char *arg;

    while (i < argc && starts_with(argv[i], "-")) {
        arg = argv[i++];
        
        if (!strcmp(HOST_OPTION, arg)) {
            if (i < argc) {
                args.host = argv[i++];
            }
            else {
                printf("%s vyžaduje IP adresu pro naslouchání, použita výchozí hodnota\n", HOST_OPTION);
            }
        }
        else if (!strcmp(PORT_OPTION, arg)) {
            int port = 0;
            
            if (i < argc) {
                if (is_integer(argv[i])) {
                    port = strtol(argv[i], NULL, 10);
                    
                    if (port < MIN_PORT || port > MAX_PORT) {
                        printf("%s není v rozsahu %d - %d\n",
                                PORT_OPTION, MIN_PORT, MAX_PORT);
                    }
                    else {
                        has_port = true;
                    }
                }
                else {
                    printf("%s není číslo\n", PORT_OPTION);
                }
                
                i++;
            }
            else {
                printf("%s vyžaduje číslo portu v rozsahu %d - %d\n",
                        PORT_OPTION, MIN_PORT, MAX_PORT);
            }

            if (has_port) {
                args.port = port;
            }
            else {
                args.port = DEFAULT_PORT;
                printf("Použita výchozí hodnota %d\n", DEFAULT_PORT);
                has_port = true;
            }
        }
        else {
            printf("Neplatný parametr %s\n", arg);
        }
    }
    
    if (!has_port) {
        printf("Nebyl zadán parametr %s\n", PORT_OPTION);
        usage();
    }
    
    return args;
}

/**
 * Vypíše nápovědu ke spuštění programu.
 */
void usage() {
    const char *name = "server";
    
    printf("\n");
    printf("TCP server hry \"Piškvorky pro více hráčů\" (Verze 2.0)\n");
    printf("Seminární práce z předmětu \"Úvod do počítačových sítí\" (KIV/UPS)\n");
    printf("Autor: Petr Kozler (A13B0359P), 2017\n");
    printf("\n");
    printf("Použití:    %s [%s <host>] %s <port>\n",
            name, HOST_OPTION, PORT_OPTION);
    printf("Popis parametrů:\n");
    printf("    <host> ... IP adresa pro naslouchání\n");
    printf("        (výchozí hodnota: \"%s\" - naslouchání na všech IP)\n",
            DEFAULT_HOST);
    printf("    <port> ... celé číslo v rozsahu %d - %d\n",
            MIN_PORT, MAX_PORT);
    printf("Příklad:    %s %s %d\n",
            name, PORT_OPTION, DEFAULT_PORT);
    printf("\n");
    
    exit(EXIT_SUCCESS);
}

File: UpsClient/src/interaction/CmdArg.java
package interaction;

import configuration.Config;
import java.net.InetAddress;
import java.net.UnknownHostException;

/**
 * Třída CmdArg slouží ke zpracování argumentů příkazové řádky.
 * 
 * @author Petr Kozler
 */
public class CmdArg {
    
    /**
     * Číslo portu serveru.
     */
    private Integer port;
    
    /**
     * Hostname nebo adresa serveru.
     */
    private InetAddress host;
    
    /**
     * Zpracuje předané argumenty.
     * 
     * @param args argumenty příkazové řádky
     */
    public CmdArg(String[] args) {
        if (args.length < 1) {
            usage();
        }
        
        int i = 0;
        String arg;
        
        while (i < args.length && args[i].startsWith("-")) {
            arg = args[i++];
            
            switch (arg) {
                case Config.HOST_OPTION:
                    InetAddress host = null;
                    
                    if (i < args.length) {
                        try {
                            host = InetAddress.getByName(args[i++]);
                        }
                        catch (UnknownHostException e) {
                            System.out.printf("%s je neznámý hostname nebo adresa\n",
                                    Config.HOST_OPTION);
                        }
                    }
                    else {
                        System.out.printf("%s vyžaduje hostname nebo IP adresu serveru\n",
                                Config.HOST_OPTION);
                    }
                    
                    if (host != null) {
                        this.host = host;
                    }
                    else {
                        try {
                            this.host = InetAddress.getByName(Config.DEFAULT_HOST);
                            System.out.printf("Použita výchozí hodnota %s\n", Config.DEFAULT_HOST);
                        }
                        catch (UnknownHostException ex) {
                            System.out.printf("Chyba při nastavování %s na výchozí hodnotu %s\n",
                                    Config.HOST_OPTION, Config.DEFAULT_HOST);
                            
                            System.exit(1);
                        }
                    }
                    break;
                case Config.PORT_OPTION:
                    Integer port = null;
                    
                    if (i < args.length) {
                        try {
                            port = Integer.parseInt(args[i++]);
                            
                            if (port < Config.MIN_PORT || port > Config.MAX_PORT) {
                                System.out.printf("%s není v rozsahu %d - %d\n",
                                        Config.PORT_OPTION, Config.MIN_PORT, Config.MAX_PORT);
                                port = null;
                            }
                        }
                        catch (NumberFormatException e) {
                            System.out.printf("%s není číslo\n", Config.PORT_OPTION);
                        }
                    }
                    else {
                        System.out.printf("%s vyžaduje číslo portu v rozsahu %d - %d\n",
                                Config.PORT_OPTION, Config.MIN_PORT, Config.MAX_PORT);
                    }
                    
                    if (port != null) {
                        this.port = port;
                    }
                    else {
                        this.port = Config.DEFAULT_PORT;
                        System.out.printf("Použita výchozí hodnota %d\n", Config.DEFAULT_PORT);
                    }
                    break;
                default:
                    System.out.printf("Neplatný parametr %s\n", arg);
                    break;
            }
        }
        
        boolean hasParams = true;
        
        if (this.host == null) {
            hasParams = false;
            System.out.printf("Nebyl zadán parametr %s\n", Config.HOST_OPTION);
        }
        
        if (this.port == null) {
            hasParams = false;
            System.out.printf("Nebyl zadán parametr %s\n", Config.PORT_OPTION);
        }
        
        if (!hasParams) {
            usage();
        }
    }
    
    /**
     * Vrátí adresu nebo název serveru.
     * 
     * @return adresa nebo název serveru
     */
    public InetAddress getHost() {
        return host;
    }
    
    /**
     * Vrátí port serveru.
     * 
     * @return port serveru
     */
    public Integer getPort() {
        return (int) port;
    }
    
    /**
     * Vypíše nápovědu ke spuštění programu.
     */
    private void usage() {
        final String name = "Client.jar";
        
        System.out.printf("\n");
        System.out.printf("TCP klient hry \"Piškvorky pro více hráčů\" (Verze 2.0)\n");
        System.out.printf("Seminární práce z předmětu \"Úvod do počítačových sítí\" (KIV/UPS)\n");
        System.out.printf("Autor: Petr Kozler (A13B0359P), 2017\n");
        System.out.printf("\n");
        System.out.printf("Použití:    %s %s <host> %s <port>\n",
                name, Config.HOST_OPTION, Config.PORT_OPTION);
        System.out.printf("Popis parametrů:\n");
        System.out.printf("    <host> ... IP adresa nebo název serveru\n");
        System.out.printf("    <port> ... celé číslo v rozsahu %d - %d\n",
                Config.MIN_PORT, Config.MAX_PORT);
        System.out.printf("Příklad:    %s %s %s %s %d\n",
                name, Config.HOST_OPTION, Config.DEFAULT_HOST, Config.PORT_OPTION, Config.DEFAULT_PORT);
        System.out.printf("\n");
        
        System.exit(0);
    }
    
}
